# Introduction

### Loadimng the data into DataBricks and exploring it using PySpark SQL. Finding intresting isights into Indian Reastaurant orders that have been palced from 2015 to 2019. Will answer different question ranging from what types are dishes/combination of dishes are most popular to what are the peaks times to help plan staffing and create new menus to satisfy customer demand.

 - ####Which dishes are the most popular?
 - ####Which dishes are most popular fish, chicken or lamb?
 - ####Which year was most profitable?
 - ####Which month is the most profitable on average?
 - ####What time of the day is most profitable?

# Data structure analysis.

In [0]:
from pyspark.sql.functions import regexp_replace, col, date_format, to_date, year, month, desc, max, min, lit, sum, when
from pyspark.sql.functions import to_timestamp

In [0]:
df_prices = sqlContext.read.load("/FileStore/tables/restaurant_1_products_price.csv", format="csv", header=True)
df_orders = sqlContext.read.load("/FileStore/tables/restaurant_1_orders.csv", format="csv", header=True)

In [0]:
display(df_prices)

Item Name Product Price Mint Sauce 0.5 Lime Pickle 0.5 Mango Chutney 0.5 Red Sauce 0.5 Onion Chutney 0.5 Plain Papadum 0.8 Spicy Papadum 0.8 Chapati 1.95 Puree 1.95 Tandoori Roti 2.6 Plain Naan 2.6 Raitha 2.95 Keema Naan 2.95 Peshwari Naan 2.95 Pilau Rice 2.95 Green Salad 2.95 Bottle Coke 2.95 Plain Rice 2.95 Garlic Naan 2.95 Onion Naan 2.95 Paratha 2.95 Bottle Diet Coke 2.95 Kulcha Naan 2.95 Lemonade 1.5 ltr 2.95 Diet Coke 1.5 ltr 2.95 Coke 1.5 ltr 2.95 Dhansak Sauce 3.95 Onion Bhajee 3.95 Lemon Rice 3.95 French Fries 3.95 Garlic Rice 3.95 Onion Rice 3.95 Perrier Water (750ml) 3.95 Special Fried Rice 3.95 Meat Samosa 3.95 Mushroom Rice 3.95 Korma Sauce 3.95 Vegetable Rice 3.95 Egg Rice 3.95 Curry Sauce 3.95 Egg Paratha 3.95 Keema Rice 3.95 Stuffed Paratha 3.95 COBRA (330ML) 3.95 Saag Rice 3.95 Masala Sauce 3.95 Madras Sauce 3.95 Vindaloo Sauce 3.95 Bengal Salad 3.95 Still Water (750ml) 3.95 Prier Water (750ml) 3.95 Onion Bhaji 3.95 Chicken Tikka 3.95 Aloo Chaat 4.95 Tandoori Chicken (1/4) 4.95 Chicken Chaat 4.95 Prawn Puree 4.95 Lamb Tikka 4.95 Vegetable Roll 4.95 Sheek Kebab 4.95 Sheek Kehab 4.95 Vegetable Samosa 4.95 King Prawn Puree 5.95 Bhindi Bhajee 5.95 Tarka Dall 5.95 Saag Paneer 5.95 Baingan Hari Mirch 5.95 Chicken Pakora 5.95 COBRA ( LARGE ) 5.95 Royal Paneer 5.95 Muttar Paneer 5.95 Chana Masala 5.95 Bombay Aloo 5.95 Saag Aloo 5.95 Aloo Methi 5.95 Saag Bhajee 5.95 Cauliflower Bhajee 5.95 Saag Dall 5.95 Aloo Gobi 5.95 Mixed Vegetable Curry 5.95 Mushroom Bhajee 5.95 Mixed Starter 5.95 Tandoori Fish 5.95 Brinjal Bhajee 5.95 Vegetable Mysore 5.95 King Prawn Butterfly 5.95 Dall Samba 5.95 COBRA (660ML) 5.95 Chicken Chaat Main 7.9 Madras - Chicken 7.95 Madras - Lamb 7.95 Curry - Chicken 7.95 Vindaloo - Chicken 7.95 Madras - Chicken Tikka 7.95 Curry - Lamb 7.95 Curry - Chicken Tikka 7.95 Vindaloo - Chicken Tikka 7.95 Vindaloo - Lamb 7.95 Madras 7.95 Curry 7.95 Vindaloo 7.95 Korma - Chicken 8.95 Chicken Tikka Masala 8.95 Pathia - Chicken Tikka 8.95 Chicken Tikka Jalfrezi 8.95 Methi - Lamb 8.95 Tandoori Chicken (Main) 8.95 Bhuna - Chicken 8.95 Korma - Chicken Tikka 8.95 Chicken Tikka (Main) 8.95 Paneer Tikka Masala 8.95 Dupiaza - Chicken 8.95 Bhuna - Chicken Tikka 8.95 Pathia - Lamb 8.95 Chicken Karahi 8.95 Bhuna - Lamb 8.95 Dhansak - Chicken Tikka 8.95 Dhansak - Chicken 8.95 Rogon - Lamb 8.95 Chicken Tikka Pasanda 8.95 Pathia - Chicken 8.95 Lamb Tikka Masala 8.95 Lamb Tikka Jalfrezi 8.95 Lamb Karahi 8.95 Madras - Prawn 8.95 Lamb Tikka (Main) 8.95 Lamb Tikka Pasanda 8.95 Lamb Tikka Garlic 8.95 Chicken Tikka Garlic 8.95 Chicken Tikka Chilli Masala 8.95 Dupiaza - Chicken Tikka 8.95 Mushroom - Chicken 8.95 Dhansak - Lamb 8.95 Sheek Kebab (Main) 8.95 Saag - Chicken 8.95 Tandoori Chicken Masala 8.95 Saag - Chicken Tikka 8.95 Dupiaza - Lamb 8.95 Curry - Prawn 8.95 Saag - Lamb 8.95 Korma - Lamb 8.95 Mushroom - Lamb 8.95 Rogon - Chicken Tikka 8.95 Lamb Tikka Chilli Masala 8.95 Vindaloo - Prawn 8.95 Methi - Chicken Tikka 8.95 Vegetable Karahi 8.95 Rogon - Chicken 8.95 Mushroom - Chicken Tikka 8.95 Methi - Chicken 8.95 Lamb Tikka Jeera 8.95 Chicken Tikka Jeera 8.95 Bhuna 8.95 Saag 8.95 Dhansak 8.95 Pathia 8.95 Korma 8.95 Dupiaza 8.95 Methi 8.95 Rogon 8.95 Mushroom 8.95 Tandoori Chicken 8.95 Chicken Shashlick 9.95 Chicken Roshni 9.95 Chicken Balti 9.95 Lamb Balti 9.95 Butter Chicken 9.95 Lamb Hari Mirch 9.95 Chicken Chilli Garlic 9.95 Lamb Biryani 9.95 Vegetable Biryani 9.95 Chicken Rezala 9.95 Lamb Rezala 9.95 Chicken Tikka Karahi 9.95 Chicken Hari Mirch 9.95 Chicken Biryani 9.95 Lamb Shashlick 9.95 Bhuna - Prawn 9.95 Chicken Haryali 9.95 Lamb Chilli Garlic 9.95 Paneer Shaslick 9.95 Chicken Mysore 9.95 Paneer Tikka Karahi 9.95 Lamb Achar 9.95 Paneer Tikka Sizzler 9.95 Chicken Tikka Sizzler 9.95 Lal Mirch Chicken 9.95 Lamb Tikka Karahi 9.95 Prawn Karahi 9.95 Lal Mirch Lamb 9.95 Korma - Prawn 9.95 Lamb Haryali 9.95 Lamb Sylhet 9.95 Vegetable Balti 9.95 Dhansak - Prawn 9.95 Lamb Mysore 9.95 Chicken Sylhet 9.95 Lamb Tikka Sizzler 9.95 Path

In [0]:
display(df_orders)

Order Number Order Date Item Name Quantity Product Price Total products 16118 03/08/2019 20:25 Plain Papadum 2 0.8 6 16118 03/08/2019 20:25 King Prawn Balti 1 12.95 6 16118 03/08/2019 20:25 Garlic Naan 1 2.95 6 16118 03/08/2019 20:25 Mushroom Rice 1 3.95 6 16118 03/08/2019 20:25 Paneer Tikka Masala 1 8.95 6 16118 03/08/2019 20:25 Mango Chutney 1 0.5 6 16117 03/08/2019 20:17 Plain Naan 1 2.6 7 16117 03/08/2019 20:17 Mushroom Rice 1 3.95 7 16117 03/08/2019 20:17 Tandoori Chicken (1/4) 1 4.95 7 16117 03/08/2019 20:17 Vindaloo - Lamb 1 7.95 7 16117 03/08/2019 20:17 Chapati 1 1.95 7 16117 03/08/2019 20:17 Lamb Tikka 1 4.95 7 16117 03/08/2019 20:17 Saag Paneer 1 5.95 7 16116 03/08/2019 20:09 Aloo Chaat 1 4.95 5 16116 03/08/2019 20:09 Chicken Pakora 1 5.95 5 16116 03/08/2019 20:09 Chicken Chilli Garlic 1 9.95 5 16116 03/08/2019 20:09 Lamb Biryani 1 9.95 5 16116 03/08/2019 20:09 Keema Naan 1 2.95 5 16115 03/08/2019 20:01 Chicken Pakora 1 5.95 7 16115 03/08/2019 20:01 Plain Papadum 1 0.8 7 16115 03/08/2019 20:01 Mango Chutney 1 0.5 7 16115 03/08/2019 20:01 Red Sauce 1 0.5 7 16115 03/08/2019 20:01 Korma - Prawn 1 9.95 7 16115 03/08/2019 20:01 Peshwari Naan 1 2.95 7 16115 03/08/2019 20:01 Pilau Rice 1 2.95 7 16114 03/08/2019 19:44 Chicken Balti 1 9.95 2 16114 03/08/2019 19:44 Special Fried Rice 2 3.95 2 16113 03/08/2019 19:42 Chicken Tikka Masala 1 8.95 5 16113 03/08/2019 19:42 Curry - Lamb 1 7.95 5 16113 03/08/2019 19:42 Pilau Rice 1 2.95 5 16113 03/08/2019 19:42 Peshwari Naan 1 2.95 5 16113 03/08/2019 19:42 Plain Papadum 1 0.8 5 16112 03/08/2019 19:41 Plain Papadum 2 0.8 4 16112 03/08/2019 19:41 Vegetable Rice 1 3.95 4 16112 03/08/2019 19:41 Saag - Prawn 1 9.95 4 16112 03/08/2019 19:41 Tandoori Chicken Masala 1 8.95 4 16111 03/08/2019 19:29 Pilau Rice 1 2.95 4 16111 03/08/2019 19:29 Lamb Tikka Chilli Masala 1 8.95 4 16111 03/08/2019 19:29 Peshwari Naan 1 2.95 4 16111 03/08/2019 19:29 Saag Aloo 1 5.95 4 16110 03/08/2019 19:28 Aloo Gobi 1 5.95 8 16110 03/08/2019 19:28 Bhindi Bhajee 1 5.95 8 16110 03/08/2019 19:28 Saag Paneer 1 5.95 8 16110 03/08/2019 19:28 Mint Sauce 2 0.5 8 16110 03/08/2019 19:28 Chicken Biryani 1 9.95 8 16110 03/08/2019 19:28 Garlic Naan 1 2.95 8 16110 03/08/2019 19:28 Plain Naan 1 2.6 8 16110 03/08/2019 19:28 Tandoori Chicken Masala 1 8.95 8 16109 03/08/2019 19:26 Plain Papadum 4 0.8 7 16109 03/08/2019 19:26 Red Sauce 2 0.5 7 16109 03/08/2019 19:26 Chicken Tikka Masala 1 8.95 7 16109 03/08/2019 19:26 Madras - Chicken Tikka 1 7.95 7 16109 03/08/2019 19:26 Mushroom Rice 2 3.95 7 16109 03/08/2019 19:26 Garlic Naan 1 2.95 7 16109 03/08/2019 19:26 Bombay Aloo 1 5.95 7 16108 03/08/2019 19:25 Onion Bhajee 1 3.95 15 16108 03/08/2019 19:25 Sheek Kebab 1 4.95 15 16108 03/08/2019 19:25 Plain Papadum 8 0.8 15 16108 03/08/2019 19:25 Mango Chutney 2 0.5 15 16108 03/08/2019 19:25 Mint Sauce 2 0.5 15 16108 03/08/2019 19:25 Onion Chutney 1 0.5 15 16108 03/08/2019 19:25 Chicken Tikka (Main) 1 8.95 15 16108 03/08/2019 19:25 Korma - Chicken 1 8.95 15 16108 03/08/2019 19:25 Chicken Tikka Masala 1 8.95 15 16108 03/08/2019 19:25 Plain Rice 1 2.95 15 16108 03/08/2019 19:25 Pilau Rice 1 2.95 15 16108 03/08/2019 19:25 Keema Naan 2 2.95 15 16108 03/08/2019 19:25 French Fries 1 3.95 15 16108 03/08/2019 19:25 Chicken Biryani 1 9.95 15 16108 03/08/2019 19:25 Mushroom Rice 1 3.95 15 16107 03/08/2019 19:16 Onion Bhajee 1 3.95 4 16107 03/08/2019 19:16 Korma - Chicken 1 8.95 4 16107 03/08/2019 19:16 Peshwari Naan 1 2.95 4 16107 03/08/2019 19:16 Pilau Rice 1 2.95 4 16106 03/08/2019 18:37 Chicken Tikka (Main) 1 8.95 3 16106 03/08/2019 18:37 Chicken Tikka Masala 1 8.95 3 16106 03/08/2019 18:37 Meat Samosa 1 3.95 3 16105 03/08/2019 18:36 Plain Papadum 2 0.8 9 16105 03/08/2019 18:36 Mango Chutney 1 0.5 9 16105 03/08/2019 18:36 Mint Sauce 1 0.5 9 16105 03/08/2019 18:36 Red Sauce 1 0.5 9 16105 03/08/2019 18:36 Keema Rice 1 3.95 9 16105 03/08/2019 18:36 Keema Naan 1 2.95 9 16105 03/08/2019 18:36 Chicken Chilli Garlic 1 9.95 9 16105 03/08/2019 18:36 Lamb Hari 

In [0]:
row_count = df_orders.count()
column_count = len(df_orders.columns)

print("Number of rows", row_count)
print("Number of columns", column_count)

Number of rows 74818
Number of columns 6


# Data Cleaning

In [0]:
df_orders.dtypes

Out[75]: [('Order Number', 'string'),
 ('Order Date', 'string'),
 ('Item Name', 'string'),
 ('Quantity', 'string'),
 ('Product Price', 'string'),
 ('Total products', 'string')]

In [0]:
df_orders = df_orders.withColumnRenamed("Order Number", "order_nr").withColumnRenamed("Order Date", "order_date").withColumnRenamed("Item Name", "item_name").withColumnRenamed("Quantity", "quantity").withColumnRenamed("Product Price", "product_price").withColumnRenamed("Total products", "total_products")

In [0]:
# As we can see we need to "order_date" column is in string format. I'm splitting it into date 
# and time column and casting them to appropriate formats.

df_orders = df_orders.withColumn("order_date", regexp_replace(col("order_date"), "/", "-"))

df_orders = df_orders.withColumn("date_time", to_timestamp(col("order_date"), "dd-MM-yyyy HH:mm"))
df_orders = df_orders.withColumn("order_date", to_date(col("date_time")))
df_orders = df_orders.withColumn("order_time", date_format(col("date_time"), "HH:mm:ss"))

In [0]:
df_orders = df_orders.withColumn("product_price", col("product_price").cast("float"))
df_orders.display()

order_nr order_date item_name quantity product_price total_products date_time order_time 16118 2019-08-03 Plain Papadum 2 0.8 6 2019-08-03T20:25:00.000+0000 20:25:00 16118 2019-08-03 King Prawn Balti 1 12.95 6 2019-08-03T20:25:00.000+0000 20:25:00 16118 2019-08-03 Garlic Naan 1 2.95 6 2019-08-03T20:25:00.000+0000 20:25:00 16118 2019-08-03 Mushroom Rice 1 3.95 6 2019-08-03T20:25:00.000+0000 20:25:00 16118 2019-08-03 Paneer Tikka Masala 1 8.95 6 2019-08-03T20:25:00.000+0000 20:25:00 16118 2019-08-03 Mango Chutney 1 0.5 6 2019-08-03T20:25:00.000+0000 20:25:00 16117 2019-08-03 Plain Naan 1 2.6 7 2019-08-03T20:17:00.000+0000 20:17:00 16117 2019-08-03 Mushroom Rice 1 3.95 7 2019-08-03T20:17:00.000+0000 20:17:00 16117 2019-08-03 Tandoori Chicken (1/4) 1 4.95 7 2019-08-03T20:17:00.000+0000 20:17:00 16117 2019-08-03 Vindaloo - Lamb 1 7.95 7 2019-08-03T20:17:00.000+0000 20:17:00 16117 2019-08-03 Chapati 1 1.95 7 2019-08-03T20:17:00.000+0000 20:17:00 16117 2019-08-03 Lamb Tikka 1 4.95 7 2019-08-03T20:17:00.000+0000 20:17:00 16117 2019-08-03 Saag Paneer 1 5.95 7 2019-08-03T20:17:00.000+0000 20:17:00 16116 2019-08-03 Aloo Chaat 1 4.95 5 2019-08-03T20:09:00.000+0000 20:09:00 16116 2019-08-03 Chicken Pakora 1 5.95 5 2019-08-03T20:09:00.000+0000 20:09:00 16116 2019-08-03 Chicken Chilli Garlic 1 9.95 5 2019-08-03T20:09:00.000+0000 20:09:00 16116 2019-08-03 Lamb Biryani 1 9.95 5 2019-08-03T20:09:00.000+0000 20:09:00 16116 2019-08-03 Keema Naan 1 2.95 5 2019-08-03T20:09:00.000+0000 20:09:00 16115 2019-08-03 Chicken Pakora 1 5.95 7 2019-08-03T20:01:00.000+0000 20:01:00 16115 2019-08-03 Plain Papadum 1 0.8 7 2019-08-03T20:01:00.000+0000 20:01:00 16115 2019-08-03 Mango Chutney 1 0.5 7 2019-08-03T20:01:00.000+0000 20:01:00 16115 2019-08-03 Red Sauce 1 0.5 7 2019-08-03T20:01:00.000+0000 20:01:00 16115 2019-08-03 Korma - Prawn 1 9.95 7 2019-08-03T20:01:00.000+0000 20:01:00 16115 2019-08-03 Peshwari Naan 1 2.95 7 2019-08-03T20:01:00.000+0000 20:01:00 16115 2019-08-03 Pilau Rice 1 2.95 7 2019-08-03T20:01:00.000+0000 20:01:00 16114 2019-08-03 Chicken Balti 1 9.95 2 2019-08-03T19:44:00.000+0000 19:44:00 16114 2019-08-03 Special Fried Rice 2 3.95 2 2019-08-03T19:44:00.000+0000 19:44:00 16113 2019-08-03 Chicken Tikka Masala 1 8.95 5 2019-08-03T19:42:00.000+0000 19:42:00 16113 2019-08-03 Curry - Lamb 1 7.95 5 2019-08-03T19:42:00.000+0000 19:42:00 16113 2019-08-03 Pilau Rice 1 2.95 5 2019-08-03T19:42:00.000+0000 19:42:00 16113 2019-08-03 Peshwari Naan 1 2.95 5 2019-08-03T19:42:00.000+0000 19:42:00 16113 2019-08-03 Plain Papadum 1 0.8 5 2019-08-03T19:42:00.000+0000 19:42:00 16112 2019-08-03 Plain Papadum 2 0.8 4 2019-08-03T19:41:00.000+0000 19:41:00 16112 2019-08-03 Vegetable Rice 1 3.95 4 2019-08-03T19:41:00.000+0000 19:41:00 16112 2019-08-03 Saag - Prawn 1 9.95 4 2019-08-03T19:41:00.000+0000 19:41:00 16112 2019-08-03 Tandoori Chicken Masala 1 8.95 4 2019-08-03T19:41:00.000+0000 19:41:00 16111 2019-08-03 Pilau Rice 1 2.95 4 2019-08-03T19:29:00.000+0000 19:29:00 16111 2019-08-03 Lamb Tikka Chilli Masala 1 8.95 4 2019-08-03T19:29:00.000+0000 19:29:00 16111 2019-08-03 Peshwari Naan 1 2.95 4 2019-08-03T19:29:00.000+0000 19:29:00 16111 2019-08-03 Saag Aloo 1 5.95 4 2019-08-03T19:29:00.000+0000 19:29:00 16110 2019-08-03 Aloo Gobi 1 5.95 8 2019-08-03T19:28:00.000+0000 19:28:00 16110 2019-08-03 Bhindi Bhajee 1 5.95 8 2019-08-03T19:28:00.000+0000 19:28:00 16110 2019-08-03 Saag Paneer 1 5.95 8 2019-08-03T19:28:00.000+0000 19:28:00 16110 2019-08-03 Mint Sauce 2 0.5 8 2019-08-03T19:28:00.000+0000 19:28:00 16110 2019-08-03 Chicken Biryani 1 9.95 8 2019-08-03T19:28:00.000+0000 19:28:00 16110 2019-08-03 Garlic Naan 1 2.95 8 2019-08-03T19:28:00.000+0000 19:28:00 16110 2019-08-03 Plain Naan 1 2.6 8 2019-08-03T19:28:00.000+0000 19:28:00 16110 2019-08-03 Tandoori Chicken Masala 1 8.95 8 2019-08-03T19:28:00.000+0000 19:28:00 16109 2019-08-03 Plain Papadum 4 0.8 7 2019-08-03T19:26:00.000+0000 19:26:00 16109 2019-08-03 Red Sauce 2 0.5 7 2019-08-03T19:26:00.000+0000 19:26:00 16109 

In [0]:
columns_to_cast = ["order_nr", "quantity", "total_products"]

df_orders = df_orders.select(col("order_date"), col("order_time"), col("product_price"), col("item_name"),
    *[col(column_name).cast("int").alias(column_name) for column_name in columns_to_cast]
)

In [0]:
df_orders.dtypes

Out[80]: [('order_date', 'date'),
 ('order_time', 'string'),
 ('product_price', 'float'),
 ('item_name', 'string'),
 ('order_nr', 'int'),
 ('quantity', 'int'),
 ('total_products', 'int')]

In [0]:
df_with_nulls = df_orders.select([col(column_name).isNull().cast("integer") for column_name in df_orders.columns])
sum_df = df_with_nulls.agg(*[sum(col).alias(col) for col in df_with_nulls.columns]).display()


CAST((order_date IS NULL) AS INT) CAST((order_time IS NULL) AS INT) CAST((product_price IS NULL) AS INT) CAST((item_name IS NULL) AS INT) CAST((order_nr IS NULL) AS INT) CAST((quantity IS NULL) AS INT) CAST((total_products IS NULL) AS INT) 0 0 0 0 0 0 0

# Data Analysis

##Which dishes are the most popular?

In [0]:
main_dishes = df_orders.filter((col("product_price") >= 3.95) & ~(col("item_name").like("%ltr%")))

top_dishes = main_dishes.groupBy(col("item_name")).sum("quantity").withColumnRenamed("sum(quantity)", "total_quantity")

top_dishes.orderBy(desc("total_quantity")).limit(5).display()



item_name total_quantity Onion Bhajee 2749 Chicken Tikka Masala 2473 Bombay Aloo 1831 Mushroom Rice 1615 Meat Samosa 1411

Databricks visualization. Run in Databricks to view.

##Which dishes are most popular lamb, fish or chicken?

In [0]:
# Dishes with Chicken
chicken_df = main_dishes.filter(col("item_name").like("%Chicken%")).select(col("item_name")).distinct().count()

# Dishes with Lamb
lamb_df = main_dishes.filter(col("item_name").like("%Lamb%")).select(col("item_name")).distinct().count()

# Dishes with Fish
fish_df = main_dishes.filter(col("item_name").like("%Fish%")).select(col("item_name")).distinct().count()


total_chicken = df_orders.filter(col("item_name").like("%Chicken%")).count()
total_lamb = df_orders.filter(col("item_name").like("%Lamb%")).count()
total_fish = df_orders.filter(col("item_name").like("%Fish%")).count()


print(f"There are {chicken_df} Chicken dishes in the menu and total {total_chicken} of ordered.")
print(f"There are {lamb_df} Lamb dishes in the menu and total of {total_lamb} ordered.")
print(f"There are {fish_df} Fish dishes in the menu and total of {total_fish} ordered.")

There are 59 Chicken dishes in the menu and total 14122 of ordered.
There are 42 Lamb dishes in the menu and total of 3757 ordered.
There are 5 Fish dishes in the menu and total of 313 ordered.


##Which year was the most profitable?

In [0]:
distinct_orders = df_orders.select("order_nr", "order_date").distinct()

# Group by year and calculate the total orders per year
order_by_year = distinct_orders.groupBy(year(col("order_date")).alias("order_date")).count()

# I have decided not to use 2015 and 2019 as the orders from these years are not complete
order_by_year.filter((col("order_date") > 2015) & (col("order_date") < 2019)).orderBy(desc("count")).display()

order_date count 2018 4553 2017 3673 2016 1872

Databricks visualization. Run in Databricks to view.

##Which month is the most profitable?

In [0]:
orders_by_month = df_orders.groupBy(month(col("order_date")).alias("order_month")).count()
orders_by_month = orders_by_month.orderBy(desc("count")).display()

order_month count 5 7202 12 7005 7 6961 4 6885 3 6773 6 6728 10 5819 11 5803 1 5713 9 5557 8 5233 2 5139

Databricks visualization. Run in Databricks to view.

##What time of the day is most profitable?

In [0]:

#Leaving only unique orders, we are going to be grouping by "time" column
distinct = df_orders.dropDuplicates(["order_nr"])

morning = (col("order_time") >= "06:00:00") & (col("order_time") < "12:00:00")
afternoon = (col("order_time") >= "12:00:00") & (col("order_time") < "18:00:00")
evening = (col("order_time") >= "18:00:00") & (col("order_time") < "22:00:00")
before_midnight = (col("order_time") >= "22:00:00") & (col("order_time") < "24:00:00")
after_midnight = (col("order_time") >= "00:00:00") & (col("order_time") < "06:00:00")

# Creating variables which are meeting conditions for a specif time of the day
df_time_of_the_day = distinct.withColumn("time_of_the_day",
                                   when(morning, "morning")
                                  .when(afternoon, "afternoon")
                                  .when(evening, "evening")
                                  .when(before_midnight, "night")
                                  .when(after_midnight, "night")
                                   .otherwise("Value if condition is not met"))


df_time_of_the_day.groupBy("time_of_the_day").count().display()

time_of_the_day count afternoon 3327 night 426 morning 151 evening 9493

Databricks visualization. Run in Databricks to view.

#Conclusion

 - ##Which dishes are the most popular?
   ### The most popular dishes Onion Bhajee and closely follwed by Chicken Tikka Masala
   
 - ##Which dishes are most popular fish, chicken lamb?
   ### There are 59 Chicken dishes in the menu and total 14122 of ordered, compared to 42 Lamb dishes in the menu and total of 3757 ordered and finally there are 5 Fish dishes in the menu and total of 313 ordered. As we can see, predominantly people prefer Chicken dishes, some people prefer Lamb and very few order fish.
   
 - ##Which year was most profitable?
   ### Most profitable year so far was 2018, byt we can clearly see that amount of orders is steadily increasing and it already doubled amount of orders from 2016. 
   
 - ##Which month is the most profitable on average?
   ### The most profitable months fall into late spring and early summer streching into mid-summer. Amount of orders sharply falls approaching the autumn and peaks again in December.
 
 - ##What time of the day is most profitable?
   ### Most orders are placed in evening between 6pm and 10pm. Although it's the shortest period it the most busiest during the day. In comparison there're 65% less orders in the morning and 95% less orders at night. Mornings orders are marginal.